crossbar/worker/proxy.py

```diff
@@ -60,6 +57,284 @@ class ProxyWorkerProcess(worker.NativeWorkerProcess):
     LOGNAME = 'Proxy'
 
 
+class ProxySession(object):
+    """
+    A router-side proxy session that handles incoming client
+    connections.
+    """
+
+    def __init__(self, router_factory):
+        self.transport = None
+        self._router_factory = router_factory
+
+        # basic session information
+        self._pending_session_id = None
+        self._realm = None
+        self._session_id = None
+        self._session_roles = None
+        self._session_details = None
+
+        # session authentication information
+        self._pending_auth = None
+        self._authid = None
+        self._authrole = None
+        self._authmethod = None
+        self._authprovider = None
+        self._authextra = None
+        self._controller = router_factory._proxy_controller
+
+        # if we have a backend connection, it'll be here (and be a
+        # Session instance)
+        self._backend_session = None
+
+        # in _transport_info are (possibly) the following keys (gleaned from router/session.py):
+        # client_cert
+        # channel_id
+        # type (websocket, etc)
+        # protocol (websocket subprotocol?) None for rawsocket
+        # peer
+        # http_headers_received
+        # http_headers_sent
+        # http_response_lines
+        # websocket_extensions_in_use
+        # cbtid
+
+        # for rawsocket, it's type, protocol=None and peer
+
+    def onOpen(self, transport):
+        """
+        Callback fired when transport is open. May run asynchronously. The transport
+        is considered running and is_open() would return true, as soon as this callback
+        has completed successfully.
+
+        :param transport: The WAMP transport.
+        :type transport: object implementing :class:`autobahn.wamp.interfaces.ITransport`
+        """
+        # print("ProxySession.onOpen: {}".format(transport))
+        self.transport = transport
+
+    from functools import singledispatchmethod
```

Comment:
oh, this is new to me: what's that? what does it do on onMessage?

Reply:
See the `_hello_received` which is decorated with `@onMessage.register(message.Hello)` .. basically now `onMessage` will (possibly) call different functions depending on the **type** of the first argument. So if a `Hello` arrives, `_hello_received` is called; for anything without a "special" message, `onMessage` itself runs.

Basically a replacement for the `if isinstance(first_arg, SomeClass):  .. elif .. elif` pattern.

Reply:
(When wrote that I expected to have more special-cases than just Hello ;) so could now simplify it to not use those decorators..?)

Reply:
interesting! thanks for explaining. however, yes, in this particular case at least, yeah. I'd prefer a basic/dumb if/else style. in the situation here, using the magic decorator IMO doesn't bring benefits, but obfuscates things.

hiding magic behind decorators is cool - sometimes. but for singledispatchmethod, I am not really convinced. because it distributes decision logic and hides it behind a magic decorator. 

taken to extreme, eg consider

```python
@arithmetic('add', a, b=3):
def my_add(self, a):
    pass

@arithmetic('mul', a, b=3):
def my_mul(self, a):
    pass
`
@arithmetic():
def my_any(self, op, args*):
    pass
`` 

awesome, beautiful, isn't it? ;)

nope. this now hides dynamic (run-time) message dispatching (is it add or mul or what) and computation behind a decorator - which is defined somewhere else

Reply:
okay, took it out